components/scream/src/share/atm_process/atmosphere_diagnostic.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef SCREAM_ATMOSPHERE_DIAGNOSTIC_HPP
+#define SCREAM_ATMOSPHERE_DIAGNOSTIC_HPP
+
+#include "share/atm_process/atmosphere_process.hpp"
+
+namespace scream
+{
+
+/*
+ * A class representing an atmosphere diagnostic.
+ * Diagnostics are similar to atmosphere processes in that
+ * they take field(s) as input and produce some output.
+ * 
+ * This subclass definition makes a few distinctions that are
+ * unique to diagnostics.
+ * 
+ * 1) Fields from the field manager can only be accessed as "Required".
+ *    Meaning that an atmosphere diagnostic can not compute or update
+ *    a field in the field manager.
+ *    - note, if you want to update something in the field manager it
+ *            should happen in an atmosphere process.
+ * 2) Diagnostics produce a single field output.
+ *    Typically the field output will be to buffered memory, meaning it
+ *    could be overwritten later in the simulation.
+ *    A diagnostic output is meant to be used for OUTPUT or a PROPERTY CHECK.
+ */
+
+class AtmosphereDiagnostic : public AtmosphereProcess
+{
+public:
+
+  // Constructor(s)
+  explicit AtmosphereDiagnostic (const ekat::Comm& comm, const ekat::ParameterList& params);
+
+  virtual ~AtmosphereDiagnostic () = default;
+
+  // The type of subcomponent
+  AtmosphereProcessType type () const { return AtmosphereProcessType::Diagnostic; }
+
+  // Getting the diagnostic output
+  Field& get_diagnostic (const Real dt);
+
+protected:
+
+  void set_computed_field_impl (const Field& f);
```

Comment:
Unfortunately, since these methods are virtual in the base class, they can still be overridden in the derived classes. You can prevent that, by adding the keyword `final` right before the semicolon.

HOWEVER, the base class also declares `set_computed_xyz` as virtual, even though most atm procs don't need to override those (notable exception is the atm proc group class). So even if you put `final` here, they may still override those methods.

The best solution is to override the `set_computed_xzy` methods (not the `_impl` ones), and mark them with `final`.